### PR TITLE
Multiple fixes for k8s forwarder

### PR DESCRIPTION
**Pull request by awly**
1. init session uploader in kubernetes service startup code
    without this, session upload directory is not getting created and causes all recordings to fail

2. move the caching layer from then entire `clusterSession` to just the ephemeral user certs
    generating user certs on the fly is the expensive part we must cache
    the rest of `clusterSession` has some fields that require extra handling for eviction: for example remote cluster or `kubernetes_service` tunnels can disappear; caching the entire `clusterSession` has little benefit performance-wise

3. use process context for emitting audit events, not request context
    request context can get cancelled by client disconnecting, losing us session.end events

4. clean up the code a bit, mostly get rid of all the embedding

Updates https://github.com/gravitational/teleport/issues/5014

### CODE REVIEW

File: lib/service/kubernetes.go [hunk 1]
```diff
@@ -81,6 +81,12 @@ func (process *TeleportProcess) initKubernetesService(log *logrus.Entry, conn *C
 		return trace.Wrap(err)
 	}
 
+	// Start uploader that will scan a path on disk and upload completed
+	// sessions to the Auth Server.
+	if err := process.initUploaderService(accessPoint, conn.Client); err != nil {
```
Comment (a-palchikov) [inline lib/service/kubernetes.go:86]: Has this been moved here from somewhere or was it just missing and it was an error/omission?

**Review by a-palchikov — APPROVED** [review]

File: lib/kube/proxy/forwarder.go [hunk 2]
```diff
@@ -115,13 +114,13 @@ type ForwarderConfig struct {
 
 // CheckAndSetDefaults checks and sets default values
 func (f *ForwarderConfig) CheckAndSetDefaults() error {
-	if f.Client == nil {
+	if f.AuthClient == nil {
 		return trace.BadParameter("missing parameter Client")
 	}
-	if f.AccessPoint == nil {
+	if f.CachingAuthClient == nil {
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/kube/proxy/forwarder.go:120]: done

File: lib/kube/proxy/forwarder.go [hunk 3]
```diff
@@ -1526,15 +1469,82 @@ func (f *Forwarder) getOrCreateRequestContext(key string) (context.Context, cont
...
 
+func (f *Forwarder) getOrRequestClientCreds(ctx authContext) (*tls.Config, error) {
+	c := f.getClientCred(ctx)
+	if c == nil {
+		return f.serializedRequestClientCreds(ctx)
+	}
+	return c, nil
+}
+
+func (f *Forwarder) getClientCred(ctx authContext) *tls.Config {
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/kube/proxy/forwarder.go]: done

File: lib/kube/proxy/forwarder.go [hunk 4]
```diff
@@ -1526,15 +1469,82 @@ func (f *Forwarder) getOrCreateRequestContext(key string) (context.Context, cont
...
+		return nil
+	}
+	c := creds.(*tls.Config)
+	if !validClientCreds(f.cfg.Clock, c) {
+		return nil
+	}
+	return c
+}
+
+func (f *Forwarder) saveClientCred(ctx authContext, c *tls.Config) error {
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/kube/proxy/forwarder.go]: done

File: lib/service/kubernetes.go (hunk 1 above)
Comment (awly) [reply lib/service/kubernetes.go:86]: It was missing here due to omission

**Review by r0mant — APPROVED** [review]

**Pull request merged**